File: chapter-1/ruby/1.1.rb
require 'test/unit/assertions'
include Test::Unit::Assertions

#Implement an algorithm to determine if a string has all unique characters.
#What if you can not use additional data structures?

def unique_characters(str)
  first_character = str[0]
  index = 1
  size = str.size

  while index < size do
    if str[index] != first_character
      return false
    end

    index = index + 1
  end

  true
end

assert unique_characters('aaaaa')
assert !unique_characters('beta')

File: chapter-1/js/1.1.js
/*
Implement an algorithm to determine if a string has all unique characters.
What if you can not use additional data structures?
*/

// Complexity = O(N)
var uniqueCharacters = function(str) {
  var firstCharacter = str[0];

  for (var i=1, size=str.length; i < size; i++) {
    if (firstCharacter != str[i]) { return false; } 
  }

  return true;
};


var assert = require('assert');
assert.equal(uniqueCharacters('aaaaa'), true);
assert.equal(uniqueCharacters('beta'), false);
